core/frontend/src/main/scala/io/udash/core/Definitions.scala

```diff
@@ -31,74 +31,90 @@ trait Presenter[S <: State] {
+trait ContainerViewFactory[S <: ContainerState[_]] extends ViewFactory[S] {
+  override def create(): (ContainerView, Presenter[S])
+}
+
+/** [[ViewFactory]] creating [[FinalView]]. */
+trait FinalViewFactory[S <: FinalState[_]] extends ViewFactory[S]{
+  override def create(): (FinalView, Presenter[S])
+}
+
 /**
-  * Abstract view which should be used in order to implement View for [[io.udash.core.ViewPresenter]].
+  * Abstract view which should be used in order to implement View for [[io.udash.core.ViewFactory]].
   * The View implementation usually gets the model and the [[io.udash.core.Presenter]] as constructor arguments.
   */
 trait View {
   /**
-    * Will be invoked by [[io.udash.routing.RoutingEngine]] in order to render the child view inside
-    * the parent view. <br/><br/>
-    *
-    * <b>This method can receive null as "view" argument, then previous child view should be removed.</b>
-    *
-    * @param view view which origins from child
-    */
-  def renderChild(view: View): Unit
-
-  /**
     * Implementation of this method should return DOM representation of view.
     * @return DOM representation of view
     */
   def getTemplate: Modifier[Element]
 
-  def apply() = getTemplate
+  @deprecated("Use `getTemplate` instead.", "0.6.0")
+  def apply(): Modifier[Element] = getTemplate
 }
 
-/** A [[io.udash.core.View]] which does not have any child view. */
-trait FinalView extends View {
-  override def renderChild(view: View): Unit =
-    if (view != null) throw View.UnexpectedChildView
-}
+/** A [[io.udash.core.View]] which can render child view. */
+trait ContainerView extends View {
+  import scalatags.JsDom.all.div
 
-object View {
-  case object UnexpectedChildView extends RuntimeException("This view is not expected to render child view. Did you forget to implement renderChild method?")
+  /** Default implementation renders child views inside this element. */
+  protected val childViewContainer: Element = div().render
+
+  /**
+    * Will be invoked by [[io.udash.routing.RoutingEngine]] in order to render the child view inside
+    * the parent view. <br/><br/>
+    *
+    * <b>This method can receive `None` as "view" argument, then previous child view should be removed.</b>
+    *
+    * @param view view which origins from child
+    */
+  def renderChild(view: Option[View]): Unit = {
+    while (childViewContainer.childElementCount > 0) {
+      childViewContainer.removeChild(childViewContainer.firstChild)
+    }
+    view.foreach(_.getTemplate.applyTo(childViewContainer))
+  }
 }
 
-/**
-  * The class which should be used to present the state for [[io.udash.routing.RoutingEngine]].
-  */
-trait State {
-  def parentState: State
+/** A [[io.udash.core.View]] which does not have any child view. */
+trait FinalView extends View
+
+/** The class which should be used to present the state for [[io.udash.routing.RoutingEngine]]. */
+trait State[HierarchyRoot <: State[HierarchyRoot]] {
```

Comment:
Maybe we should use a type member instead? See: https://github.com/ghik/opinionated-scala/wiki/Generics-and-type-members#type-member-solution 
It would be easier to migrate existing code, we could get nice benefits from expressing some of the relations introduced by this PR with dependent types and there would be less underscores messing with generics inference.